include/gatekeeper_lls.h

```diff
@@ -19,6 +19,181 @@
 #ifndef _GATEKEEPER_LLS_H_
 #define _GATEKEEPER_LLS_H_
 
-int run_lls(void);
+#include <rte_timer.h>
+
+#include "gatekeeper_mailbox.h"
+#include "gatekeeper_net.h"
+
+/*
+ * Maximum key length (in bytes) for an LLS map. It should be set
+ * to the maximum key length needed. Currently, this is set to 16 for
+ * the number of bytes in an IPv6 address.
+ */
+#define LLS_MAX_KEY_LEN (16)
+
+/* Number of records that a LLS cache can hold. */
+#define LLS_CACHE_RECORDS (1024)
+
+/* Requests that can be made to the LLS block. */
+enum lls_req_ty {
+	/* Express interest in a map by registering a callback function. */
+	LLS_REQ_HOLD,
+	/*
+	 * Remove a previously-submitted hold, if not already deleted
+	 * by virtue of a callback function signaling it should not
+	 * be invoked again.
+	 */
+	LLS_REQ_PUT,
+};
+
+/* Replies that come from the LLS block. */
+enum lls_reply_ty {
+	/* The reply represents a map resolution (or update to one). */
+	LLS_REPLY_RESOLUTION,
+	/*
+	 * The reply is a notification that the hold is
+	 * removed, so the requester can free state as needed.
+	 */
+	LLS_REPLY_FORCE,
+};
+
+/* Map that is returned to blocks that request resolution. */
+struct lls_map {
+	/* Ethernet address of this map. */
+	struct ether_addr ha;
+
+	/* Port on which this map exists. */
+	uint8_t           port_id;
+
+	/* Whether this map has been marked as stale. */
+	int               stale;
+
+	/* IP address for this map, in network ordering. */
+	uint8_t           ip_be[LLS_MAX_KEY_LEN];
+};
+
+/*
+ * Format of callback function for requesting LLS maps.
+ *
+ * The LLS block invokes the callbacks, so each block should ensure
+ * that the callback function deals with any race conditions and
+ * is aware that the blocks may reside in different NUMA nodes.
+ *
+ * If the requesting block wants future updates, it should set
+ * *@call_again to true before returning. Otherwise, by default
+ * the LLS block will remove the interest from the block.
+ */
+typedef void (*lls_req_cb)(const struct lls_map *map, void *arg,
+	enum lls_reply_ty ty, int *call_again);
+
+/* A hold for an LLS map. */
+struct lls_hold {
+	/* Cache that holds (or will hold) this map. */
+	struct lls_cache *cache;
```

Comment:
I see -- we need to separate what's only needed for the request and what's needed to be stored in the record.